src/lib/notify.js

```diff
@@ -0,0 +1,221 @@
+import tabs from 'sdk/tabs';
+import { getMostRecentBrowserWindow } from 'sdk/window/utils';
+
+import Logger from './log';
+import sendEvent from './metrics';
+
+const logger = new Logger(
+  'sdk.lib.notify',
+  getMostRecentBrowserWindow().console
+);
+
+class BaseElement {
+  getWindow() {
+    return Services.wm.getMostRecentWindow('navigator:browser');
+  }
+}
+
+class RatingStar extends BaseElement {
+  constructor(val, cb) {
+    super(...arguments);
+    this.element = this.render(val, cb);
+  }
+
+  handleClick(evt, cb) {
+    cb(evt.target.getAttribute('data-score'));
+  }
+
+  render(val, cb) {
+    const win = this.getWindow();
+    const elem = win.document.createElement('toolbarbutton');
+    elem.classList.add('plain', 'star-x');
+    elem.id = `star-${val}`;
+    elem.setAttribute('data-score', val);
+    elem.addEventListener('click', evt => this.handleClick(evt, cb));
+    return this.style(elem);
+  }
+
+  style(elem) {
+    elem.style.height = '20px';
+    elem.style.width = '20px';
+    return elem;
+  }
+}
+
+class Rating extends BaseElement {
+  constructor(cb) {
+    super(...arguments);
+    this.element = this.render(cb);
+  }
+
+  render(cb, max = 5) {
+    const { document } = this.getWindow();
+    const fragment = document.createDocumentFragment();
+    const elem = document.createElement('hbox');
+    elem.id = 'star-rating-container';
+    for (let i = 0; i < max; i++) {
+      elem.appendChild(new RatingStar(max - i, cb).element);
+    }
+    fragment.appendChild(this.style(elem));
+    fragment.appendChild(this.renderSpacer());
+    return fragment;
+  }
+
+  renderSpacer() {
+    const { document } = this.getWindow();
+    const elem = document.createElement('spacer');
+    elem.flex = 66;
+    return elem;
+  }
+
+  style(elem) {
+    elem.style.marginBottom = '2px';
+    elem.style.mozBoxDirection = 'reverse';
+    elem.style.setProperty('-moz-box-direction', 'reverse', 'important');
+    return elem;
+  }
+}
+
+export default class Notification extends BaseElement {
+  constructor(options) {
+    super(...arguments);
+    this.id = uuid();
+    this.type = 'random';
+
+    this.notifyBox = null;
+    this.getSiteName().then(sitename => {
+      options.label = `How would you rate your experience on ${sitename}?`;
+      this.element = this.render(options);
+    });
+  }
+
+  promptedPing() {
+    const promptedPing = {
+      method: 'pulse-prompted',
+      id: this.id,
+      type: this.type
+    };
+    logger.log('Prompted', promptedPing);
+    sendEvent(promptedPing);
+  }
+
+  dismissedPing() {
+    const dismissedPing = {
+      method: 'pulse-dismissed',
+      id: this.id,
+      type: this.type
+    };
+    logger.log('Dismissed', dismissedPing);
+    sendEvent(dismissedPing);
+  }
+
+  dismiss(notifyBox, notification) {
+    notifyBox.removeNotification(notification);
+  }
+
+  openSurvey(surveyUrl, sentiment) {
+    storage.id[this.id] = tabs.activeTab;
```

Comment:
Finally starting to get a look today. FWIW, I saw an error from around here:
```
TypeError: _simpleStorage.storage.id is undefined
```
Which is weird, because [it looks like that should be initialized](https://github.com/chuckharmston/pulse/blob/da853d9269a066abdec207546bf350fe7093ec3f/src/index.js#L36). Going to keep poking and see if it was a fluke